tests/shiplift_client.rs

```diff
@@ -76,3 +79,50 @@ async fn should_wait_for_at_least_one_second_before_fetching_logs_shiplift() {
     assert_that(&(after_run - before_run)).is_greater_than(Duration::from_secs(1));
     assert_that(&(after_logs - before_logs)).is_less_than(Duration::from_secs(1));
 }
+
+#[tokio::test(threaded_scheduler)]
+async fn shiplift_run_command_should_include_env_and_volumes() {
+    let mut volumes = HashMap::new();
+    volumes.insert("/tmp".to_owned(), "/hostmp".to_owned());
+
+    let mut env_vars = HashMap::new();
+    env_vars.insert("one-key".to_owned(), "one-value".to_owned());
+    env_vars.insert("two-key".to_owned(), "two-value".to_owned());
+
+    let image = HelloWorld { volumes, env_vars };
+    let run_args = RunArgs::default();
+
+    let docker = clients::Shiplift::new();
+    let container = docker.run_with_args(image, run_args).await;
+
+    // inspect volume and env
+    let container_details = docker
+        .client
+        .containers()
+        .get(container.id())
+        .inspect()
+        .await
+        .unwrap();
+
+    let envs = container_details.config.env.unwrap();
+    assert_that!(&envs).contains(&"one-key=one-value".into());
+    assert_that!(&envs).contains(&"two-key=two-value".into());
+}
+
+#[tokio::test(threaded_scheduler)]
+async fn shiplift_run_command_should_expose_all_ports_if_no_explicit_mapping_requested() {
+    let image = HelloWorld::default();
+    let docker = clients::Shiplift::new();
+    let container = docker.run(image).await;
+
+    // inspect volume and env
+    let container_details = docker
+        .client
+        .containers()
+        .get(container.id())
+        .inspect()
+        .await
+        .unwrap();
+
+    assert_that!(container_details.host_config.publish_all_ports).is_equal_to(true);
+}
```

Comment:
You could move these tests into the `src/clients/shiplift.rs` file, then you have access to private fields :)

Reply:
That's a good point. I was gonna ask if there's a pattern of leaving tests actually running container out of unit test. If you have no concern I'll move them back to `shiplift.rs`

Reply:
The only difference between tests in `tests/` and tests within the module is that tests in `tests/` can only access the public API and are hence a good way of testing the lib in a way users will have to deal with it. For example:

- does the module structure make sense (in the `use` statements)?
- are all types accessible that need to be accessible?

In this case, we are testing an implementation detail of the `Shiplift` client, so it is totally fine to do that within the actual module.

I am not fussed about the "unit-test go into the module and integration-test go into `tests/" policy. In my experience, it is usually quite hard to even find an agreement on what a unit test actually is :)

Sidenote: Excellent talk on this topic: https://www.youtube.com/watch?v=EZ05e7EMOLM

Generally, it would be ideal if as many tests as possible - be it unit or integration tests - are as fast as possible. Starting containers is slow, so if you can test things in an effective way without starting containers that is great. If starting containers is necessary to ensure things actually work, that is also ok :)

Reply:
Thanks for the thorough explanation! I'm with you on unit test / integration distinction sometimes unnecessary. Rust makes it very easy to organize tests so we can just focus on getting good coverage. These tests are moved back into `shiplift.rs`.

Since the API is a wrapper around shiplift, it's hard to test its functionalities without starting a real container. These tests are running blazing fast (at least on my Mac), so shouldn't be a concern for now.  :)